File: tracking/tracking_validation.py
"""
tracking_validation.py

Validation runtime for:
- Kalman prediction
- Temporal smoothing
- Motion-aware tracking
- Persistent object identity
- Missed-frame continuity

Tracking Pipeline
-----------------
predict
    ->
associate
    ->
update
"""

import random
import time

from tracking.tracker_reference import (
    MultiObjectTracker
)


def build_detection(
    center_x,
    center_y,
    noise_std=0.0
):
    """
    Build noisy synthetic detection.
    """

    noisy_x = (
        center_x +
        random.uniform(
            -noise_std,
            noise_std
        )
    )

    noisy_y = (
        center_y +
        random.uniform(
            -noise_std,
            noise_std
        )
    )

    return {

        "center_x": noisy_x,

        "center_y": noisy_y,

        "yaw_rad": 0.0,

        "width_px": 120,

        "height_px": 80,

        "confidence": 1.0
    }


def print_tracks(
    tracks
):
    """
    Print active tracker state.
    """

    print("\nActive Tracks:\n")

    if len(tracks) == 0:

        print("No active tracks.")

        return

    for track in tracks:

        print(

            f"Track ID: {track.track_id} | "

            f"Center: "
            f"({track.center_x:.2f}, "
            f"{track.center_y:.2f}) | "

            f"Velocity: "
            f"({track.velocity_x:.2f}, "
            f"{track.velocity_y:.2f}) | "

            f"Missed: "
            f"{track.missed_frames} | "

            f"Age: "
            f"{track.age}"
        )


def main():

    print(
        "\n=== Kalman Tracking Validation ===\n"
    )

    # ------------------------------------------------------------
    # Tracker
    # ------------------------------------------------------------

    tracker = MultiObjectTracker(

        max_missed_frames=5,

        association_distance=150.0
    )

    # ------------------------------------------------------------
    # Simulated trajectory
    # ------------------------------------------------------------

    simulated_frames = [

        # --------------------------------------------------------
        # Stable motion
        # --------------------------------------------------------

        [

            build_detection(
                100,
                100,
                noise_std=8
            )
        ],

        [

            build_detection(
                112,
                108,
                noise_std=8
            )
        ],

        [

            build_detection(
                126,
                118,
                noise_std=8
            )
        ],

        [

            build_detection(
                141,
                129,
                noise_std=8
            )
        ],

        # --------------------------------------------------------
        # Temporary dropout
        # --------------------------------------------------------

        [],

        [],

        # --------------------------------------------------------
        # Reappearance
        # --------------------------------------------------------

        [

            build_detection(
                175,
                152,
                noise_std=8
            )
        ],

        [

            build_detection(
                191,
                164,
                noise_std=8
            )
        ],

        [

            build_detection(
                208,
                178,
                noise_std=8
            )
        ]
    ]

    # ------------------------------------------------------------
    # Runtime loop
    # ------------------------------------------------------------

    for frame_idx, detections in enumerate(
        simulated_frames
    ):

        print(
            "\n================================="
        )

        print(
            f"Frame {frame_idx}"
        )

        print(
            "================================="
        )

        print(
            f"\nDetections: "
            f"{len(detections)}"
        )

        # --------------------------------------------------------
        # Tracker update
        # --------------------------------------------------------

        tracks = tracker.update(
            detections
        )

        # --------------------------------------------------------
        # Print tracker state
        # --------------------------------------------------------

        print_tracks(
            tracks
        )

        # --------------------------------------------------------
        # Print temporal history
        # --------------------------------------------------------

        for track in tracks:

            print(
                f"\nTrack {track.track_id} "
                f"History Size: "
                f"{len(track.history)}"
            )

        time.sleep(0.5)

    # ------------------------------------------------------------
    # Final summary
    # ------------------------------------------------------------

    print(
        "\n=== Validation Complete ===\n"
    )

    print(
        "Expected behavior:"
    )

    print(
        "- Stable persistent IDs"
    )

    print(
        "- Reduced jitter"
    )

    print(
        "- Smooth motion continuity"
    )

    print(
        "- Track survival during dropout"
    )

    print(
        "- Motion-aware prediction"
    )


if __name__ == "__main__":
    main()

File: tracking/src/tracking_validation.cpp
/*
tracking_validation.cpp
*/

#include <iostream>

#include "tracking_validation.hpp"
#include "track_state.hpp"


void TrackingValidation::run()
{
    std::cout << "\n=== TrackState Validation ===\n";

    // ------------------------------------------------------------
    // Create track
    // ------------------------------------------------------------

    TrackState track(

        1,

        320.0f,
        240.0f,

        0.0f,

        120.0f,
        80.0f
    );

    std::cout << "\nInitial Track:\n";

    std::cout
        << "Track ID: " << track.track_id << "\n"
        << "Center: ("
        << track.center_x << ", "
        << track.center_y << ")\n"
        << "Yaw: " << track.yaw_rad << "\n"
        << "Dimensions: "
        << track.width_px << " x "
        << track.height_px << "\n";

    // ------------------------------------------------------------
    // Simulated updates
    // ------------------------------------------------------------

    float positions[4][2] = {

        {325.0f, 242.0f},
        {330.0f, 245.0f},
        {338.0f, 249.0f},
        {345.0f, 255.0f}
    };

    for (int i = 0; i < 4; i++)
    {
        track.update(

            positions[i][0],
            positions[i][1],

            0.05f * i,

            120.0f + i,
            80.0f + i
        );

        std::cout
            << "\nUpdate " << i + 1 << ":\n";

        std::cout
            << "Center: ("
            << track.center_x << ", "
            << track.center_y << ")\n";

        std::cout
            << "Velocity: ("
            << track.velocity_x << ", "
            << track.velocity_y << ")\n";

        std::cout
            << "Age: "
            << track.age << "\n";

        std::cout
            << "History Size: "
            << track.history.size() << "\n";
    }

    // ------------------------------------------------------------
    // Missed frames
    // ------------------------------------------------------------

    std::cout
        << "\nSimulating missed frames...\n";

    track.markMissed();

    track.markMissed();

    std::cout
        << "Missed Frames: "
        << track.missed_frames << "\n";

    // ------------------------------------------------------------
    // Deactivation
    // ------------------------------------------------------------

    std::cout
        << "\nDeactivating track...\n";

    track.deactivate();

    std::cout
        << "Active: "
        << track.is_active << "\n";

    std::cout
        << "\nValidation complete.\n";
}